### ISSUE TITLE
OnTouch only trigger with char movement, not npc movement.

### ISSUE THREAD

**Issue by wbgneto** [rathena/rathena#502]
Well, as title says, if I use npcwalkto, and this npc has a OnTouch event, the event won't trigger even if there are users on the touch area. It just triggers if the user walk to npc touch area.

I believe the function unit_walktoxy_timer misses a case BL_NPC on the type switch.

I'm currently trying to fix it, If I have success, I'll post here.
Thanks.

**Comment by walterneto** [comment]
Well, was very easy to fix.
Just add a case BL_NPC, with a map_foreachinarea, executing a function that runs thru every player on npc touch area. Then call npc functions that handle touch ( you can see on the BL_PC case ).

**Comment by cydh** [comment]
and how it works.
>OnTouch:

This label will be executed if a trigger area is defined for the NPC object it's 
in. If it isn't present, the execution will start from the beginning of the NPC 
code. **The RID of the triggering character object** will be attached.

**Comment by walterneto** [comment]
Ok... so what? If this is you saying "it's working as intended"? o.o

I don't know why you highlighted "The RID of the triggering character object"... let me try to explain again:

1- Make NPC with OnTouch area defined move from point A to point B.
2- Stand still with a character in the middle of the trajectory from A to B.
3- NPC will pass you, you will enter OnTouch area, but your RID won't be attached and OnTouch won't trigger.

If you merely posted the OnTouch explanation from docs just to add info, please disregard my indignation haha.

**Comment by nanakiwurtz** [comment]
So what you are trying to say is, the new coordinates of the OnTouch area doesn't get updated on new position?

**Comment by RagnarokNova** [comment]
What he's trying to say is that if a character walks into an NPC, OnTouch works fine. But if an NPC walks into a character (unitwalk or npcwalk) OnTouch does not activate.

**Comment by nanakiwurtz** [comment]
Okay, so for example the original coordinates was on 100,100 with 2x2 OnTouch area, the NPC walks to 150,150. On which area does the OnTouch label triggers?
100,100 or during the walk or 150,150

**Comment by RagnarokNova** [comment]
On 150,150. But if you were already standing in 150,150 before the NPC moved, and the NPC walked on top of you, OnTouch will not activate unless you leave the area and walk back inside it.

**Comment by RagnarokNova** [comment]
I'm pretty sure he means the OnTouch is moving with the NPC as it's walking too. It's just the activation is messed up.

**Comment by walterneto** [comment]
The triggering doesn't work. 
The OnTouch area moves with the npc, but while it moves, if new standing characters enter the areas, it won't trigger.

To make it work, I did this:
In function unit_walktoxy_timer(int tid, unsigned int tick, int id, intptr_t data), added cases BL_NPC:
http://puu.sh/iPnhF/9a4fbfc61f.png
http://puu.sh/iPnji/18a3dca329.png

and created the function 'unit_npc_walk_touch'
http://puu.sh/iPnlB/fa625def43.png
This function I simply copied the BL_PC case of the second switch

It worked, with an exception, it will only trigger on the first time it pass you, if you keep standing and the npc comes back, it won't trigger again. ( For what I want to use, this is no problem, but to fix it in rAthena, it should work ).
My solution probably isn't the best way, so I'll just leave for rA great devs :)

**Comment by firecroll** [comment]
If i recall correctly in official was done on purpose for some event npc like the speed pots that appears random in 3 places in which one of them lets you buy some. But the other 2 just stole you some zeny and dissapear to one of the 2 random places remaining. So for ppl not to stay with 3 chars with a enter macro until it appears in the required place and make really easy buying them, they obliged you to do it yourself going to each place or  moving the char or ask 2 friends to do it in order to buy quickly.
Maybe you are talking about other npc that walk by but i think that npc works in the same way. You have to pass by yourself, If the npc appears beside you it wont trigger.
It was like a countermeasure.
Again maybe my memory fails. xD

**Comment by walterneto** [comment]
Well, I don't know.
I just think it's bad for it to not work with walking npc's.
The dev team can decide whether to fix it or not, mine is working fine, so whatever.
